Gutenberg: Introduce a reusable JetpackPluginSidebar

tyxla opened the pull request:
As @jeherve suggested, if we introduce a plugin sidebar for Jetpack, it should be reusable and any block / extension we create should be able to add to it separately and independently.

#### Changes proposed in this Pull Request

* Introduce a reusable Jetpack sidebar as the `<JetpackPluginSidebar />` component.
* Make usage of `<JetpackPluginSidebar />` add a Jetpack item in the plugin sidebar more menu.
* Make usage of `<JetpackPluginSidebar />` add a Jetpack item in the toolbar.

#### Preview

Refer to #28899 for previews.

#### Testing instructions

* Not directly testable, use #28899 for testing.

tyxla commented on client/gutenberg/extensions/presets/jetpack/editor-shared/jetpack-plugin-sidebar.scss:
```diff
@@ -0,0 +1,40 @@
+		&.is-toggled:hover {
+			.jetpack-logo,
+			.jetpack-logo * {
+				fill: none !important;
```
While it's generally bad to have `!important` in any styles, this is required, as otherwise Gutenberg will override the colors inside our SVG.

ockham replied:
The entire selector is kinda awful :stuck_out_tongue_closed_eyes: 

Specifically, do we need this `fill` override? In my element inspector, it looks like it's being overridden by https://github.com/Automattic/wp-calypso/pull/28913/files#diff-b3f6b257d3217c67ad46a47c0acf88bbR32 anyway.

sirreal replied:
I've landed https://github.com/Automattic/wp-calypso/pull/28920 so I expect we can drop this `*` selector bit at least.

tyxla replied:
I've removed the unnecessary `fill` in 82aec101c8a1e3d92e0ab3da06b75905211ea988 and removed the `*` selector in favor of concrete selectors in 3d529479713f76bb4228c63fd8ce5fe99a767a72 - we still need to remove the stroke that Gutenberg adds.

tyxla commented on client/gutenberg/extensions/presets/jetpack/editor-shared/jetpack-plugin-sidebar.js:
```diff
@@ -0,0 +1,41 @@
+	</Fill>
+);
+
+JetpackPluginSidebar.Slot = () => (
```
The `Fill` / `Slot` functionality could be non-straightforward. Dev who're looking at this, I'd appreciate it if you tell me what was unclear in this file, so I could add some inline documentation.

ockham replied:
I'm not too familiar with the pattern, but it seems to be a standard util in Gutenberg. You could add a code comment to https://wordpress.org/gutenberg/handbook/components/slot-fill/, but OTOH, where do we draw the line when it comes to adding explicit links to WP docs? So maybe don't add a link :slightly_smiling_face:

tyxla replied:
Thanks for verifying folks 👍

ockham commented on client/gutenberg/extensions/presets/jetpack/editor-shared/jetpack-plugin-sidebar.js:
```diff
@@ -0,0 +1,41 @@
+
+const { Fill, Slot } = createSlotFill( 'JetpackPluginSidebar' );
+
+const JetpackIcon = () => <JetpackLogo />;
```
I don't quite get why we need this indirection? Why not use `<JetpackLogo />` inline where this PR is now using `<JetpackIcon />`?

tyxla replied:
Good call - that's a remnant from something else I had there. Removed in e02e6248aa11979164f4ef0e8a3883171836b9fe.

ockham commented on client/gutenberg/extensions/presets/jetpack/editor-shared/jetpack-plugin-sidebar.scss:
```diff
@@ -0,0 +1,40 @@
+			}
+
+			.jetpack-logo__icon-triangle {
+				fill: #fff !important;
```
Should we use a named color here?

sirreal replied:
https://github.com/Automattic/wp-calypso/blob/d0a2b47e1935a2388783621e1ccd5a8e9fe62102/assets/stylesheets/shared/_colors.scss#L52-L53
https://github.com/Automattic/color-studio/blob/302bab1a6279ce539cd47cf369bf307a0d300d71/dist/colors.scss#L3

tyxla replied:
👍 e9dfd4e10f1f8c0107652830fcd201b2aa6ac6a5

sirreal commented on client/gutenberg/extensions/presets/jetpack/editor-shared/jetpack-plugin-sidebar.scss:
```diff
@@ -0,0 +1,40 @@
+			}
+
+			.jetpack-logo__icon-circle {
+				fill: #00be28 !important;
```
https://github.com/Automattic/wp-calypso/blob/d0a2b47e1935a2388783621e1ccd5a8e9fe62102/assets/stylesheets/shared/_colors.scss#L16-L17

tyxla replied:
👍 e9dfd4e10f1f8c0107652830fcd201b2aa6ac6a5

tyxla commented:
Thanks for the thorough feedback, folks ❤️ 

I've addressed it all - care to have another look? 

If you wish to test - I've also updated #28899 and cherry-picked everything there.

sirreal commented on client/gutenberg/extensions/presets/jetpack/editor-shared/jetpack-plugin-sidebar.scss:
```diff
@@ -0,0 +1,40 @@
+			}
+
+			.jetpack-logo__icon-triangle {
+				fill: $muriel-white !important;
```
I think we want just `$white`. Our main shared colors assign `$white: $muriel-white`. I don't think we should rely on that variable here. Sorry if my other comment was unclear.

tyxla replied:
Well, they both are `#fff` now, so 🤷‍♂️ . Updated in 8bd78e4805cae2f3a847223dd690c29cca95dfcb

simison commented on client/gutenberg/extensions/presets/jetpack/editor-shared/jetpack-plugin-sidebar.js:
```diff
@@ -0,0 +1,39 @@
+		<PluginSidebarMoreMenuItem target="jetpack" icon={ <JetpackLogo /> }>
+			{ __( 'Jetpack' ) }
+		</PluginSidebarMoreMenuItem>
+		<PluginSidebar name="jetpack" title={ __( 'Jetpack' ) } icon={ <JetpackLogo /> }>
```
Mini-mini nit and totally late to this, but I don't think we translate `Jetpack` elsewhere.

tyxla replied:
Oh, that's right. Good catch 👍 

Fixed in #28988
